pkg/master/reconcilers/reconcilers.go

```diff
@@ -0,0 +1,55 @@
+/*
+Copyright 2017 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+// Package reconcilers Endpoint Reconcilers for the apiserver
+package reconcilers
+
+import (
+	"net"
+
+	"k8s.io/kubernetes/pkg/api"
+)
+
+// EndpointReconciler knows how to reconcile the endpoints for the apiserver service.
+type EndpointReconciler interface {
+	// ReconcileEndpoints sets the endpoints for the given apiserver service (ro or rw).
+	// ReconcileEndpoints expects that the endpoints objects it manages will all be
+	// managed only by ReconcileEndpoints; therefore, to understand this, you need only
+	// understand the requirements.
+	//
+	// Requirements:
+	//  * All apiservers MUST use the same ports for their {rw, ro} services.
+	//  * All apiservers MUST use ReconcileEndpoints and only ReconcileEndpoints to manage the
+	//      endpoints for their {rw, ro} services.
+	//  * ReconcileEndpoints is called periodically from all apiservers.
+	ReconcileEndpoints(serviceName string, ip net.IP, endpointPorts []api.EndpointPort, reconcilePorts bool) error
+}
+
+const (
+	// MasterCountReconcilerType will select the original reconciler
+	MasterCountReconcilerType = "master-count"
```

Comment:
Let's make a wrapper type for this since these are the only values. That'll make it clearer than 'string' when referencing the type.

Specifically something like:

```go
type ReconcilerType string

const (
   MasterCountReconcilerType ReconcilerType = "master-count"
   LeaseEndpointReconcilerType = "lease"
   NoneEndpointReconcilerType = "none"
)
```

And from there, we can also change the type name to just `Type` since it's already in a `reconciler` package; `reconcilers.Type` vs `reconcilers.ReconcilerType`.

Reply:
I think it adds too much boilerplate to convert from the named type to a string (ie, for the strings.Join() call) and elsewhere in the code. The code would need to be instrumented to cast from the type to a string and vice versa.

Golang 1.9 has aliases, perhaps it could be tweaked in the future to be a golang type alias?

Reply:
Type aliases aren't meant for this iiuc, so I don't think that helps.

For the sake of `AllTypes`, I think using a `[]string` for that seems fine. I can't think of other places, besides the flags, where there should be any boilerplate?

Alternately, you could make `[]ReconcilerType` have a `.String()` method which joins them by commas and have the flags print `AllTypes.String()`

Reply:
Fixed, though I named the function 'Names' instead of forcing a format